### PR TITLE
added basic queue implementation and tests.

### PR DESCRIPTION
**Pull request by chrismeyers**

### CODE REVIEW

File: queue/queue.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	return len(q.vals) == 0
+}
+
+func (q *GenericQueue) Push(val Generic) {
+	q.vals = append(q.vals, val)
+}
+
+func (q *GenericQueue) Pop() Generic {
+	if q.IsEmpty() {
+		return nil
```
Comment (mattmeyers) [inline queue/queue.go]: Return the Zero value here. It's going to change based on the type.

File: queue/queue.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	}
+
+	front := q.vals[0]
+	copy(q.vals, q.vals[1:])
```
Comment (mattmeyers) [inline queue/queue.go]: This will panic when popping the last element.

File: queue/queue.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+import (
+	"fmt"
+	"strings"
+)
+
+// Generic represents a generic value.
+type Generic interface{}
+
+// Zero is the default zero value for a queue.
+var Zero Generic = nil
```
Comment (mattmeyers) [inline queue/queue.go]: This isn't being used and can be removed.

File: queue/queue.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+func newDNode(data Generic) *dNode {
+	return &dNode{data: data}
+}
+
+func (q *GenericQueue) String() string {
+	if q.IsEmpty() {
+		return "[]"
+	}
+
+	sb := strings.Builder{}
```
Comment (mattmeyers) [inline queue/queue.go:45]: I updated how linked lists are printed in order to match slice printing. Update this to match that. There shouldn't be blank spaces between the brackets and elements.

File: queue/queue.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+// This method will panic when attempting to pop an empty queue.
+func (q *GenericQueue) Pop() Generic {
+	if q.IsEmpty() {
+		panic("queue is empty")
+	}
+
+	n := q.head
+	q.len--
+
+	if n.prev == nil {
```
Comment (mattmeyers) [inline queue/queue.go]: This logic is specific to removing an arbitrary node in the list and can be optimized for removing just the head. All we need is to move the head pointer to the next node and change that node's prev pointer.

**Comment by chrismeyers** [comment]
I think that https://github.com/chrismeyers/armory/blob/queue/queue/queue.go#L99-L101 and https://github.com/chrismeyers/armory/blob/queue/queue/queue.go#L105-L107 are unreachable because you can only pop the head of a queue.  Would you agree?

**Review by mattmeyers — COMMENTED** [review]
Most of that code is unreachable. For the entire pop operation you should only need:
```go
q.head = n.next
if q.head != nil {
  q.head.prev = nil
}
```
Double check my logic. I might be missing something.

**Comment by mattmeyers** [comment]
Turns out my logic is slightly off. Need to account for the tail when popping the last element.
```
q.head = n.next
if q.head != nil {
  q.head.prev = nil
} else {
  q.tail = nil
}
```

**Pull request merged**